script.py

```diff
@@ -0,0 +1,130 @@
+"URL info caching script."
+
+import csv
+import json
+import os
+import sys
+from urllib.parse import urlparse
+import requests
+
+OBJECT_SPECS = {
+    'checkouts': {
+        'patchset_files': [{'url': True}],
+    },
+    'builds': {
+        'input_files': [{'url': True}],
+        'output_files': [{'url': True}],
+        'config_url': True,
+        'log_url': True,
+    },
+    'tests': {
+        'output_files': [{'url': True}],
+        'log_url': True,
+    },
+}
+
+
+def extract_fields(spec, data):
+    """
+    Extract values of fields from data according to a specification.
+
+    Args:
+        spec:   The specification of fields to extract the values from.
+        data:   The data to extract the field values from.
+
+    Returns:
+        An array of tuples, where each tuple contains:
+            * A tuple containing dictionary keys leading to the field
+            * The field value
+    """
+    tuples = []
+
+    if isinstance(spec, dict):
+        for key, value in spec.items():
+            if key in data:
+                for ext_tuple in extract_fields(value, data[key]):
+                    tuples.append(((key,) + ext_tuple[0], ext_tuple[1]))
+    elif isinstance(spec, list):
+        assert len(spec) == 1
+        for item in enumerate(data):
+            tuples += extract_fields(spec[0], item)
+    elif spec is True:
+        return [(tuple(), data)]
+
+    return tuples
```

Comment:
At this point it's completely nit picking from my part, but how do you feel about

- moving the recursion's base case to the top.
- repeating the variable names (obj_type, obj_spec and obj) you used in the other spec iteration (inside main)
- using more variables and comments trying to break down the logic
- I guess both conditions need to check the type of `data` as well.

```suggestion
    if spec is True:
        return [(tuple(), data)]

    if isinstance(spec, dict) and isinstance(data, dict):
        for obj_type, obj_spec in spec.items():
            if obj := data.get(obj_type):
                inner_tuples = extract_fields(obj_spec, obj)

                # prefix the path from inner tuples with obj_type
                for field_path, field_value in inner_tuples:
                    tuples.append(((obj_type,) + field_path, field_value))

    elif isinstance(spec, list) and isinstance(data, list):
        assert len(spec) == 1
        obj_spec = spec[0]
        for obj in data:
            tuples += extract_fields(obj_spec, obj)

    return tuples
```

Reply:
> I guess both conditions need to check the type of data as well.

The motivation for that is to avoid trying to do `.get` on a list...
We could alternatively raise an error when `data` type does not match `spec` 🤷🏼

Reply:
Okay